lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -133,14 +133,44 @@ def safe_change_column_default(table_name, column_name, default_value)
   end
 
   def safe_make_column_nullable(table, column)
+    quoted_table_name = connection.quote_table_name(table)
+    quoted_column_name = connection.quote_column_name(column)
+
     safely_acquire_lock_for_table(table) do
-      raw_execute "ALTER TABLE #{table} ALTER COLUMN #{column} DROP NOT NULL"
+      raw_execute "ALTER TABLE #{quoted_table_name} ALTER COLUMN #{quoted_column_name} DROP NOT NULL"
     end
   end
 
+  # Postgres 12+ can use a valid CHECK constraint to proove that no values of a column are null, avoiding
+  # locking the table when altering a column to NOT NULL
+  #
+  # Source:
+  # https://dba.stackexchange.com/questions/267947/how-can-i-set-a-column-to-not-null-without-locking-the-table-during-a-table-scan/268128#268128
+  # (https://archive.is/X55up)
+  def safe_make_column_not_nullable(table, column)
+    if ActiveRecord::Base.connection.postgresql_version < 12_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Cannot safely make a column non-nullable before Postgres 12"
+    end
+
+    quoted_table_name = connection.quote_table_name(table)
+    quoted_column_name = connection.quote_column_name(column)
+
+    tmp_constraint_name = :tmp_not_null_constraint
```

Comment:
Are constraint names global? If so, do we want to add a random tag on the end of this?

Reply:
Yeah, I think constraint names need to be unique within a schema. Because this method is doing multiple things non-transactionally, there is a possibility we could error out before the constraint is removed. So a) I agree with @jcoleman that we should add some random characters to this string and b) we should document that this might leave behind a constraint on failure, similar to creating a [partitioned index](https://github.com/braintree/pg_ha_migrations?tab=readme-ov-file#safe_add_concurrent_partitioned_index).

Reply:
https://www.postgresql.org/docs/current/sql-set-constraints.html#:~:text=Because%20PostgreSQL%20does%20not%20require,will%20act%20on%20all%20matches.